[fio extras] storage: allow target to be marked as bad

detsch opened the pull request:
This feature is required to allow a rollback operation to be forced on a target whose installation was finished.

mike-sul commented on src/libaktualizr/storage/sqlstorage.cc:
```diff
@@ -1126,7 +1137,7 @@ void SQLStorage::saveInstalledVersion(const std::string& ecu_serial, const Uptan
         target.correlation_id(), static_cast<int>(update_mode == InstalledVersionUpdateMode::kCurrent),
         static_cast<int>(update_mode == InstalledVersionUpdateMode::kPending),
-        static_cast<int>(update_mode == InstalledVersionUpdateMode::kCurrent || old_was_installed), old_id.value());
+        static_cast<int>(update_mode == InstalledVersionUpdateMode::kCurrent /*|| old_was_installed*/), old_id.value());
```
Why do we need to comment `|| old_was_installed`?

detsch replied:
I left this by mistake, from a previous test. I'll revert it.

mike-sul commented on src/libaktualizr/storage/invstorage.h:
```diff
@@ -116,8 +116,8 @@ class INvStorage {
-  virtual bool loadInstallationLog(const std::string& ecu_serial, std::vector<Uptane::Target>* log,
-                                   bool only_installed) const = 0;
+  virtual bool loadInstallationLog(const std::string& ecu_serial, std::vector<Uptane::Target>* log, bool only_installed,
+                                   bool allow_current = true) const = 0;
```
I would break this change into two commits, one for the "allow_current" change and the second one for marking target as a "bad/failing" target.

mike-sul replied:
Probably we should call the variable as `include_current`.

detsch replied:
`include_current` sounds good.

detsch committed:
[fio extras] storage: allow target to be marked as bad

This feature is required to allow a rollback operation to be forced on
a target whose installation was finished.

Signed-off-by: Andre Detsch <andre.detsch@foundries.io>

detsch committed:
[fio extras] storage: allow removing current version from installed list

An additional option is included to allow filtering out current target
from installed versions list.

This is required to properly select the a rollback target after
installation was finalized, and there was not ostree update, only apps.

Signed-off-by: Andre Detsch <andre.detsch@foundries.io>

mike-sul commented on src/libaktualizr/storage/sqlstorage.cc:
```diff
@@ -1118,6 +1118,17 @@ void SQLStorage::saveInstalledVersion(const std::string& ecu_serial, const Uptan
       LOG_ERROR << "Failed to save installed versions: " << db.errmsg();
       return;
     }
+  } else if (update_mode == InstalledVersionUpdateMode::kBadTarget) {
```
At first glance, it looks like this `if` is better suited in the `if else` below before `if (!!old_id)`.
```
if (update_mode == InstalledVersionUpdateMode::kBadTarget) {
...
} else if (!!old_id) {
...
```
What do you think?

detsch replied:
Sorry, missed this comment. It makes sense, to avoid doing 2 updates. I'm testing the change and will update the PR accordingly.

mike-sul commented on src/libaktualizr/package_manager/ostreemanager.cc:
```diff
@@ -395,7 +395,13 @@ Uptane::Target OstreeManager::getCurrent() const {
   storage_->loadPrimaryInstalledVersions(&current_version, nullptr);
 
-  if (!!current_version && current_version->sha256Hash() == current_hash) {
+  bool staged_reboot_pending = false;
```
Nice, thanks for this fix.

I think we should move all required functionality from the aktualizr's ostreemanager into the aklite's rootfsmanager going further in order to get rid of dependency on the package manager interface at all as well as on aktualizr in general. Not related to this PR, just strategic direction.

detsch replied:
This change seems to be break aklite unit tests:
```
The following tests FAILED:
        114 - test_aklite_rollback (Failed)
        115 - test_aklite_rollback_ext (Failed)
        119 - test_aklite_offline (Failed)
        121 - test_cli (Failed)
```
I need to take a better look to understand the reason.

mike-sul replied:
Because those tests are adjusted for "getCurrent" returning invalid current target in non-booted environment.

We need either:

a) adjust these tests to the change;
b) revert this change, and adjust the e2e test somehow so we can get desired behavior in the e2e test environment. Maybe we can play with `ostree admin deploy` command or something like that to make ostreemanager change its opinion about current target :).

detsch replied:
I've dropped the change from this PR for now, so we can merge the user rollback feature regardless of it.
